### ISSUE TITLE
Diagnostic message disappearing

### ISSUE THREAD

**Issue by 0xfMel** [rust-lang/rust-analyzer#18194]
**rust-analyzer version**: rust-analyzer 0.0.0 (03456c545 2024-08-22)

**rustc version**: rustc 1.82.0-beta.4 (8c27a2ba6 2024-09-21)

**editor or extension**: Neovim v0.10.1

Commit 03456c545f6f8abd507f9ee16b85f17d5d9346ee introduced an issue where diagnostics disappear.  Diagnostics are visible when first opening a file, but saving any file in the project will cause diagnostic messages to disappear from all files **for an affected crate**.  After making some changes to the crate, I also had the issue occur without saving the file.  The diagnostics just flash up on the editor quickly before disappearing.  The changes I made were just commenting out some unused code and fixing a compilation error.

This only seems to affect diagnostic messages from `cargo check`, things such a missing semicolons that are detected on-the-fly will still show up and remain through saves.

Unfortunately I've been unable to produce a minimal reproducible example, and cannot share my codebase.  Hopefully knowing the commit will help though.

The issue is in a workspace with 10 crates, and seemingly only affects one of them.  Nothing about the affected crate stands out as different to the others in any significant way.  The crate has no other errors or warnings from cargo check, other than dead code which I suppressed with `#![allow(dead_code)]` (removing the suppression has no effect on this issue).

My :LspLog is empty other than the following

```
[START][2024-09-25 22:02:32] LSP logging initiated
[WARN][2024-09-25 22:02:32] .../lua/vim/lsp.lua:628	"buf_attach_client called on unloaded buffer (id: 4): "
```

However this is also shown for previous, working versions and doesn't seem to be causing any issues.

All other language server features appear to be normal (go-tos, code actions, auto-complete).

**Comment by 0xfMel** [comment]
As is typical after creating an issue, I've figured out what the problem with the crate was.
One of the other crates in the workspace had a dependency on tokio with the fs feature enabled, the crate which was experiencing this issue was using the fs the feature but I hadn't added this feature to my dependency on tokio in the workspace Cargo.toml.

As a result, running `cargo check` inside the workspace directory was able to ignore the fs feature being missing inside this crate, but running cargo check inside of the crate's directory instead would cause it to only show the error about trying to use fs without the feature enabled.

So presumably this commit changed the behavior of where the diagnostics for a crate within a workspace come from.

This seems like some very unintuitive behavior on the part of cargo's dependency resolution, but I presume its working as intended.
Wouldn't know what you want to do about this, but it seems like a RA bug to me.

This can very easily be reproduced with the following:

Cargo.toml
```toml
[workspace]
members = [
	"depcrate",
	"issuecrate",
]
resolver = "2"

[workspace.dependencies]
tokio = "1.40.0"
```

depcrate/Cargo.toml
```toml
[package]
name = "depcrate"
version = "0.1.0"
edition = "2021"

[dependencies]
tokio = { version = "1.40.0", features = ["fs"] }
```

issuecrate/Cargo.toml
```toml
[package]
name = "issuecrate"
version = "0.1.0"
edition = "2021"

[dependencies]
tokio = { workspace = true }
```

issuecrate/main.rs
```rust
use tokio::fs::File;

fn main() {
    File::open("/test")
}
```

When first opening `issuecrate/main.rs`, the diagnostics suggest adding a semicolon to avoid returning the file future.
After the first save, the diagnostic changes to an unresolved import, "unable to import `tokio::fs`, could not find `fs` in `tokio`"
